dss/index/es/document.py

```diff
@@ -208,48 +212,70 @@ def _prepare_index(self, dryrun):
-                schema_version_list.append((key_name, schema_info.version))
+            schema = SchemaInfo.from_json(file_content)
+            if schema is not None:
+                if file_name.endswith('_json'):
+                    file_name = file_name[:-5]
+                # Enforce the prerequisites that make the mapping to shape descriptors bijective. This will enable us
+                # to parse shape descriptors should we need to in the future. Dots have to be avoided because they
+                # are used as separators. A number (the schema version) is used to terminate each file's entry in the
+                # shape descriptor, allowing us to distinguishing between the normal form of an entry and the
+                # compressed form used when schema and file name are the same.
+                reject('.' in file_name, f"A metadata file name must not contain '.' characters: {file_name}")
+                reject(file_name.isdecimal(), f"A metadata file name must contain at least one non-digit: {file_name}")
+                reject('.' in schema.type, f"A schema name must not contain '.' characters: {schema.type}")
+                reject(schema.type.isdecimal(), f"A schema name must contain at least one non-digit: {schema.type}")
+                assert '.' not in schema.version, f"A schema version must not contain '.' characters: {schema.version}"
+                assert schema.version.isdecimal(), f"A schema version must consist of digits only: {schema.version}"
+                schemas_by_file[file_name] = schema
             else:
                 logger.warning(f"Unable to obtain JSON schema info from file '{file_name}'. The file will be indexed "
                                f"as is, without sanitization. This may prevent subsequent, valid files from being "
                                f"indexed correctly.")
-
-        if schema_version_list:
-            schema_version_list = sorted(schema_version_list, key=lambda e: e[0])
-            if 1 == len(set(e[1] for e in schema_version_list)):
-                return 'v' + schema_version_list[0][1]
+        if schemas_by_file:
+            same_version = 1 == len(set(schema.version for schema in schemas_by_file.values()))
+            same_schema_and_file_name = all(file_name == schema.type for file_name, schema in schemas_by_file.items())
+            if same_version and same_schema_and_file_name:
+                return 'v' + next(iter(schemas_by_file.values())).version
             else:
-                return 'v-' + '-'.join([e[0] + e[1] for e in schema_version_list])
+                schemas = sorted(schemas_by_file.items())
+
+                def entry(file_name, schema):
+                    if schema.type == file_name:
+                        return file_name, schema.version
+                    else:
+                        return file_name, schema.type, schema.version
+
+                return '.'.join(chain('v', *(starmap(entry, schemas))))
```

Comment:
This may amount to personal preference but I find `'v.' + '.'.join([entry(*t) for t in schemas])` to be more readable, and it avoids the itertools imports.

Reply:
I'll try it. It does look more readable.

Generally speaking, though, I don't see why itertools is problematic. It can lead to more readable and more optimal code.